packages/workbox-background-sync/lib/Queue.mjs

```diff
@@ -0,0 +1,297 @@
+      successfullyReplayedRequests.push({request, response});
+
+      // Remove the request from IndexedDB asynchronously (don't await).
+      // TODO(philipwalton): in the unlikely event that the delete fails,
+      // this request may be replayed again. Do we want to handle this case?
+      this._removeRequest(key);
+    }
+
+    if (allReplaysSuccessful) {
+      this._runCallback('allRequestsDidReplay', successfullyReplayedRequests);
+    } else {
+      this._registerSync();
+    }
+  }
+
+  /**
+   * Gets all requests in the object store matching this queue's name.
+   *
+   * @private
+   * @return {Promise<Array>}
+   */
+  async _getStorableRequestsInQueue() {
+    const db = await this._getDb();
+    const storableRequests = [];
+
+    for (const [key, entry] of (await db.getAll()).entries()) {
+      if (entry.queueName == this._name) {
+        // Requests older than `maxRetentionTime` should be ignored.
+        const storableRequest = new StorableRequest(entry.storableRequest);
+
+        if (Date.now() - storableRequest.timestamp > this._maxRetentionTime) {
+          // No need to await this since it can happen in parallel.
+          this._removeRequest(key);
+          continue;
+        }
+
+        storableRequests.push([key, storableRequest]);
+      }
+    }
+
+    return storableRequests;
+  }
+
+  /**
+   * Gets a reference to the IndexedDB object store for queued requests.
+   *
+   * @private
+   * @return {Promise<DBWrapper>}
+   */
+  _getDb() {
+    return indexedDBHelper.getDB(
+        DB_NAME, OBJECT_STORE_NAME, {autoIncrement: true});
+  }
+
+  /**
+   * Removes a request from IndexedDB for the specified key.
+   *
+   * @private
+   * @param {string} key
+   */
+  async _removeRequest(key) {
+    const db = await this._getDb();
+    await db.delete(key);
+  }
+
+  /**
+   * Runs the passed callback if it exists.
+   *
+   * @private
+   * @param {string} name The name of the callback on this._callbacks.
+   * @param {...*} args The arguments to invoke the callback with.
+   */
+  _runCallback(name, ...args) {
+    if (typeof this._callbacks[name] == 'function') {
+      this._callbacks[name].apply(null, args);
+    }
+  }
+
+  /**
+   * In sync-supporting browsers, this adds a listener for the sync event.
+   * In non-sync-supporting browsers, this will retry the queue on service
+   * worker startup.
+   *
+   * @private
+   */
+  _addSyncListener() {
+    self.addEventListener('sync', (event) => {
+      event.waitUntil(this.replayRequests());
+    });
+
+    // If the browser doesn't support background sync, retry
+    // every time the service worker starts up as a fallback.
+    if (!('sync' in registration)) {
+      this.replayRequests();
+    }
+  }
+
+  /**
+   * Registers a sync event with a tag unique to this instance.
+   *
+   * @private
+   */
+  async _registerSync() {
+    try {
+      await this._waitUntilActive();
+      await registration.sync.register(`${TAG_PREFIX}:${this._name}`);
```

Comment:
If the queue name is unique, why are we adding a prefix? We can just use the queue name. Would be what the developer probably expects.